File: pc02/pc02_06/cpp/pc02_06.cpp
/**
 * Erdos numbers
 * pc110206
 * uva10044
 */

#include <iostream>
#include <map>
#include <vector>
#include <queue>
#include <string>
#include <sstream>
using namespace std;

vector<int> getAuthors(const string& paper, map<string, int>& authorIndex, int& nextIndex) {
	vector<int> authorIds;
	size_t colonPos = paper.find(':');
	string authorPart = (colonPos != string::npos) ? paper.substr(0, colonPos) : paper;

	stringstream ss(authorPart);
	string part;

	vector<string> parts;
	while (getline(ss, part, ',')) {
		size_t start = part.find_first_not_of(" \t");
		size_t end = part.find_last_not_of(" \t");
		if (start != string::npos) {
			parts.push_back(part.substr(start, end - start + 1));
		}
	}

	for (size_t i = 0; i + 1 < parts.size(); i += 2) {
		string author = parts[i] + ", " + parts[i + 1];

		if (authorIndex.find(author) == authorIndex.end()) {
			authorIndex[author] = nextIndex++;
		}
		authorIds.push_back(authorIndex[author]);
	}

	return authorIds;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);

	int T;
	cin >> T;
	cin.ignore();

	bool isFirst = true;

	while (T-- > 0) {
		if (!isFirst) cout << '\n';
		isFirst = false;

		cout << "Scenario " << (T + 1) << '\n';

		int P, N;
		cin >> P >> N;
		cin.ignore();

		map<string, int> authorIndex;
		int nextIndex = 0;

		vector<vector<int>> graph(10000);

		string line;
		for (int i = 0; i < P; i++) {
			getline(cin, line);
			vector<int> authorIds = getAuthors(line, authorIndex, nextIndex);

			for (size_t j = 0; j < authorIds.size(); j++) {
				for (size_t k = j + 1; k < authorIds.size(); k++) {
					graph[authorIds[j]].push_back(authorIds[k]);
					graph[authorIds[k]].push_back(authorIds[j]);
				}
			}
		}

		vector<int> dist(nextIndex, -1);

		if (authorIndex.find("Erdos, P.") != authorIndex.end()) {
			int erdosId = authorIndex["Erdos, P."];
			queue<int> q;
			q.push(erdosId);
			dist[erdosId] = 0;

			while (!q.empty()) {
				int u = q.front();
				q.pop();

				for (int v : graph[u]) {
					if (dist[v] == -1) {
						dist[v] = dist[u] + 1;
						q.push(v);
					}
				}
			}
		}

		for (int i = 0; i < N; i++) {
			getline(cin, line);

			if (authorIndex.find(line) == authorIndex.end() || dist[authorIndex[line]] == -1) {
				cout << line << " infinity\n";
			} else {
				cout << line << " " << dist[authorIndex[line]] << '\n';
			}
		}
	}

	return 0;
}

File: pc02/pc02_06/python/pc02_06.py
"""
Erdos numbers
pc110206
uva10044
"""

import sys
from collections import defaultdict, deque


def get_authors(paper):
    """Extract authors from paper string."""
    colon_pos = paper.find(':')
    author_part = paper[:colon_pos] if colon_pos != -1 else paper

    parts = [p.strip() for p in author_part.split(',') if p.strip()]

    authors = []
    for i in range(0, len(parts) - 1, 2):
        author = f"{parts[i]}, {parts[i + 1]}"
        authors.append(author)

    return authors


def calculate_erdos_numbers(papers, queries):
    """Calculate Erdos numbers for all authors."""
    author_index = {}
    next_index = 0
    graph = defaultdict(list)

    for paper in papers:
        authors = get_authors(paper)

        author_ids = []
        for author in authors:
            if author not in author_index:
                author_index[author] = next_index
                next_index += 1
            author_ids.append(author_index[author])

        for i in range(len(author_ids)):
            for j in range(i + 1, len(author_ids)):
                u, v = author_ids[i], author_ids[j]
                graph[u].append(v)
                graph[v].append(u)

    dist = [-1] * next_index

    if "Erdos, P." in author_index:
        erdos_id = author_index["Erdos, P."]
        queue = deque([erdos_id])
        dist[erdos_id] = 0

        while queue:
            u = queue.popleft()

            for v in graph[u]:
                if dist[v] == -1:
                    dist[v] = dist[u] + 1
                    queue.append(v)

    results = []
    for query in queries:
        if query not in author_index or dist[author_index[query]] == -1:
            results.append(f"{query} infinity")
        else:
            results.append(f"{query} {dist[author_index[query]]}")

    return results


def main():
    lines = sys.stdin.readlines()
    T = int(lines[0].strip())

    line_idx = 1
    scenario = 1

    first = True

    for _ in range(T):
        if not first:
            print()
        first = False

        print(f"Scenario {scenario}")
        scenario += 1

        P = int(lines[line_idx].strip())
        N = int(lines[line_idx + 1].strip())
        line_idx += 2

        papers = []
        for _ in range(P):
            papers.append(lines[line_idx].strip())
            line_idx += 1

        queries = []
        for _ in range(N):
            queries.append(lines[line_idx].strip())
            line_idx += 1

        results = calculate_erdos_numbers(papers, queries)
        for result in results:
            print(result)


if __name__ == '__main__':
    main()
